mart/utils/modality_dispatch.py

```diff
@@ -0,0 +1,68 @@
+#
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+from __future__ import annotations
+
+from itertools import cycle
+from typing import Any, Callable, Iterable
+
+import torch
+
+__all__ = ["modality_dispatch"]
+
+DEFAULT_MODALITY = "default"
+
+
+def modality_dispatch(
+    modality_func: Callable | dict[str, Callable],
+    data: torch.Tensor | dict[str, torch.Tensor] | Iterable[Any],
+    *,
+    input: torch.Tensor | dict[str, torch.Tensor] | Iterable[Any],
+    target: torch.Tensor | Iterable[Any] | None,
+    modality: str = DEFAULT_MODALITY,
+):
+    """Recursively dispatch data and input/target to functions of the same modality.
+
+    The function returns an object that is homomorphic to input and data.
+    """
+
+    assert type(data) == type(input)
```

Comment:
Note that this will break for universal perturbation idea, since I want to apply the same perturbation to every input. instead, you should dispatch on types below and raise an NotImplementedError if you don't support that combination of types.

Reply:
The new `singledispatch` implementation does not have the assertion.